request.js

```diff
@@ -1028,26 +1031,11 @@ Request.prototype.onRequestResponse = function (response) {
           break
       }
     }
-  } else if (response.statusCode === 401 && self._hasAuth && !self._sentAuth) {
-    var authHeader = response.caseless.get('www-authenticate')
-    var authVerb = authHeader && authHeader.split(' ')[0].toLowerCase()
-    debug('reauth', authVerb)
-
-    switch (authVerb) {
-      case 'basic':
-        self.auth(self._user, self._pass, true)
-        redirectTo = self.uri
-        break
-
-      case 'bearer':
-        self.auth(null, null, true, self._bearer)
-        redirectTo = self.uri
-        break
-
-      case 'digest':
-        self._digest(authHeader)
-        redirectTo = self.uri
-        break
+  } else if (response.statusCode === 401) {
+    var authHeader = self._auth.response(self.method, self.uri.path, response.headers)
```

Comment:
I intentionally made it that way, so that the module knows less about the consumer, and can work with more generic input data parameters.

Also in `caseless` that's just an assignment, so I don't think it degrades performance or anything like that.

```js
this.dict = dict || {}
```